### PR TITLE
Agent drain: implement agent side. (Remaining: server)

### PR DESCRIPTION
**Pull request by jkh52**
Agent drain: first half of https://github.com/kubernetes-sigs/apiserver-network-proxy/issues/566

- Agent sends drain signal to Server.
- Server only logs it.
- Test cases.

Note that if a new agent binary sends the DRAIN message to an old server binary, it is harmless and will be ignored: https://github.com/kubernetes-sigs/apiserver-network-proxy/blob/30266153048bd27662cdf878eae36efbb079941e/pkg/server/server.go#L994

### CODE REVIEW

File: cmd/agent/app/server.go [hunk 1]
```diff
@@ -92,7 +95,31 @@ func (a *Agent) Run(o *options.GrpcProxyAgentOptions, stopCh <-chan struct{}) er
...
+	go runpprof.Do(context.Background(), labels, func(context.Context) { handleSignals(c, drain, stop) })
+
+	return drain, stop
+}
+
+func handleSignals(signalCh chan os.Signal, drainCh, stopCh chan struct{}) {
+	s := <-signalCh
+	klog.V(2).InfoS("Received first signal", "signal", s)
+	close(drainCh)
+	s = <-signalCh
```
Comment (cheftako) [inline cmd/agent/app/server.go:117]: Once we begin draining, do we want to timeout and stop if no second signal is sent?

**Comment by cheftako** [comment]
To be clear this is doing nothing to actually "drain"? Its just a prep step for that later PR?

File: cmd/agent/app/server.go (hunk 1 above)
Comment (jkh52) [reply cmd/agent/app/server.go:117]: I think the most cautious (and simple) decision is no, keep an agent in this state fully functioning.

**Comment by jkh52** [comment]
> To be clear this is doing nothing to actually "drain"? Its just a prep step for that later PR?

Right, this PR should be a functional no-op for the agent, except for sending the draining hint to the server (for later PR: agent selection).

**Comment by cheftako** [comment]
/lgtm
Looking forward to the actual drain logic being added. We clearly need to discuss edge cases such as what to do when ALL agents are in the drain state.

**Pull request merged**